// fetch-component-data/lib/middleware.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = createFetchComponentDataMiddlware;
var _lodash = _interopRequireDefault(require("lodash"));
var _reactRouterConfig = require("react-router-config");
var _fetchComponentData = _interopRequireDefault(require("./fetchComponentData"));
var routeChangeTypes = ['@@reduxReactRouter/routerDidChange', '@@router/LOCATION_CHANGE'];
var locsEqual = function locsEqual(locA, locB) {
  return locA && locB && locA.pathname === locB.pathname && locA.search === locB.search;
};
function createFetchComponentDataMiddlware(getRoutes) {
  var routes = getRoutes();
  return function (store) {
    return function (next) {
      return function (action) {
        var router = store.getState().router;
        var currentLocation = router.location;
        if (currentLocation && _lodash["default"].includes(routeChangeTypes, action.type)) {
          var newLocation = action.payload && action.payload.location || action.payload;
          if (locsEqual(currentLocation, newLocation)) return;
          var branch = (0, _reactRouterConfig.matchRoutes)(routes, newLocation.pathname);
          (0, _fetchComponentData["default"])({
            store: store,
            branch: branch,
            action: action,
            location: newLocation
          });
        }
        return next(action);
      };
    };
  };
}
module.exports = exports.default;
